### ISSUE TITLE
[BUG] Object keys are not detected in all calles 

### ISSUE THREAD

**Issue by kachkaev** [francoismassart/eslint-plugin-tailwindcss#267]
**Describe the bug**

There several common `callees` which support objects with class names in keys. [The existing implementation of `tailwindcss/no-custom-classname`](https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/6b6c0dd28e123cc118bff83654f951f736fa58e8/lib/rules/no-arbitrary-value.js#L110) only scans object keys in a function named `classnames`. This results in false negatives when the rule is enabled.

https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/6b6c0dd28e123cc118bff83654f951f736fa58e8/lib/rules/no-arbitrary-value.js#L109-L116

**To Reproduce**

```js
// file=eslint.config.js
{
  "extends": ["next/core-web-vitals", "plugin:tailwindcss/recommended"],
  "settings": {
    "tailwindcss": {
      "config": "./tailwind.config.js",
      "callees": [
        "cva", // https://cva.style

        // a project is typically configured with one of these, the shape of arguments match `classnames`
        "classnames",
        "classNames",
        "clsx",
        "cn",
        "cns",
        "cx"
      ]
    }
  },
  "rules": {
    "tailwindcss/no-custom-classname": "error"
  }
}
```

```ts
import { cva } from "class-variance-authority";
import classnames from "classnames";
import Image from "next/image";

cva("text-white", "bg-black", {
  someVariant: {
    s: "p-4",
    "p-oops": "p-10", // 🎉 no ESLint error here (key is not a class name in `cva`)
    anything: "p-oops", // 🎉 ESLint error here: `Classname 'p-oops' is not a Tailwind CSS class! tailwindcss/no-custom-classname`
  },
});

classnames("text-white", "bg-black", {
  "p-4": true,
  "p-oops": true, // 🎉 ESLint error here: `Classname 'p-oops' is not a Tailwind CSS class! tailwindcss/no-custom-classname`
});

// `classNames` is an alternative import name from `classnames`
const classNames = classnames;
classNames("text-white", "bg-black", {
  "p-4": true,
  "p-oops": true, // 🚨 no ESLint error here
});

// https://www.npmjs.com/package/clsx claims to be a faster alternative to `classnames`
const clsx = classnames;
clsx("text-white", "bg-black", {
  "p-4": true,
  "p-oops": true, // 🚨 no ESLint error here
});

// `cn` is a shorthand for `classnames`, seen used as a wrapper for `clsx` + `tailwind-merge`
// https://github.com/reactjs/react.dev/blob/842c24c9aefaa60b7ae9b46b002bd1b3cf4d31f3/src/components/Tag.tsx#L5
// https://github.com/shadcn-ui/ui/blob/33a5fc7966cfe8887dac732f64582f40162869c6/apps/www/lib/utils.ts#L4-L6
const cn = classnames;
cn("text-white", "bg-black", {
  "p-4": true,
  "p-oops": true, // 🚨 no ESLint error here
});

// Some people may prefer `cns` instead of `cn` as a shorthand for `classnames`
// https://github.com/francoismassart/eslint-plugin-tailwindcss/issues/99#issuecomment-1014361250
const cns = classnames;
cns("text-white", "bg-black", {
  "p-4": true,
  "p-oops": true, // 🚨 no ESLint error here
});
```

**Expected behavior**

I’d expect all my callees except `cva` to work just a `classnames()` function. It seems more reasonable than to put `cva` in the deny-list rather than `classnames` into the allow list for class names in keys.

**Screenshots**

<img width="957" alt="code snippet from above, with squiggly lines under ESLint errors" src="https://github.com/francoismassart/eslint-plugin-tailwindcss/assets/608862/76f0ea1d-de79-4ca4-80c6-f8c04242dd28">

**Environment (please complete the following information):**
 - OS: macOS
 - Softwares + version used:
     - `"eslint-plugin-tailwindcss": "3.13.0"`

**Additional context**

Support for object keys was requested in #99 and implemented in #103. Then #135 was rased, requesting support for `cva`. This issue was resolved in #185. When doing so, support for all `classname`-like functions was dropped.

If the sole goal of #185 is to support `cva` and we don’t want to add another config option (`calleesWithClassNamesInObjectKeys` 😬), maybe this fix would do?

```ts
 case 'ObjectExpression': 
   const isUsedByCva = node.callee && node.callee.name === 'cva'; 
   const isVue = node.key && node.key.type === 'VDirectiveKey'; 
   arg.properties.forEach((prop) => { 
     const propVal = !isUsedByCva || isVue ? prop.key : prop.value; 
     parseForArbitraryValues(node, propVal); 
   }); 
   return; 
```

In the meantime, I will probably change the signature of my custom `cn` function (`clsx` + `tailwind-merge`):

```ts
/**
 * @example
 *
 *   ```ts
 *   cn(
 *     'some-class some-other-class',
 *      conditionA && 'some-additional-class',
 *      conditionB ? 'some-class-foo' : 'some-class-bar'
 *     )
 *   ```
 *
 * @param inputs A list of strings (including conditional ones)
 *
 *               We don’t use `ClassValue` from `clsx` because it also allows objects.
 *               Object keys are not checked by `eslint-plugin-tailwindcss`.
 */
export function cn(...inputs: (string | number | null | boolean | undefined)[]) {
  return mergeTailwindClasses(clsx(inputs));
}
```

**eslint config file or live demo**

...MWE is being uploaded...

**Comment by kachkaev** [comment]
I managed to patch `eslint-plugin-tailwindcss` in https://github.com/francoismassart/eslint-plugin-tailwindcss/issues/267 and it seems to be doing what I expect. Can this be converted into a PR or am I missing a use case that can throw a spanner in the works?

**Comment by kachkaev** [comment]
I’ve stopped using this patch myself. Turns out that object syntax inside `cn` is not that useful and `&&` works equally well:

```ts
cn("text-white", "bg-black", {
  "p-4": condition,
  "p-oops": anotherCondition, // 🚨 no ESLint error here
});
```
↓
```ts
cn(
  "text-white",
  "bg-black",
  condition && "p-4",
  anotherCondition && "p-oops", // 🎉 ESLint error here: `Classname 'p-oops' is not a Tailwind CSS class! tailwindcss/no-custom-classname`
});
```

My `cn` function:

```ts
/**
 * Same as `ClassNameValue` from `tailwind-merge`, but without support for arrays
 * (to keep arguments simple)
 */
export type ClassNameValue = string | null | undefined | 0 | false;

/**
 * @example `cn('foo', condition && 'bar', condition && 'baz')`
 */
export function cn(...inputs: ClassNameValue[]): string {
  return twMerge(inputs);
}
```

Because objects inside `cva` are supported by `eslint-plugin-tailwindcss`, all class names in my projects are covered even without a patch. Here is what I have in `.eslintrc.js`:

```ts
  settings: {
    tailwindcss: {
      callees: ['cn', 'cva'],
      ignoredKeys: ['defaultVariants'],
      config: `${__dirname}/tailwind.config.ts`,
    },
  },
```

This issue is still valid, but it possible to bypass it by stepping a bit to the side.

**Comment by asportnoy** [comment]
@francoismassart would you be able to add @kachkaev's patch above to this repo? Would be great to have.

**Comment by kachkaev** [comment]
@zegs21 consider dropping object arguments completely, as described in https://github.com/francoismassart/eslint-plugin-tailwindcss/issues/267#issuecomment-1907798102. I’m quite happy with just `&&`s for conditional styles. There is only one way of doing one thing instead of two 👍

**Comment by francoismassart** [comment]
Closing this issue, cleaning up before release of v4.
